gdal/alg/marching_squares/square.h

```diff
@@ -0,0 +1,425 @@
+#ifndef MARCHING_SQUARE_SQUARE_H
+#define MARCHING_SQUARE_SQUARE_H
+
+#include <cassert>
+#include <math.h>
+#include "utility.h"
+#include "point.h"
+
+namespace marching_squares {
+
+struct Square
+{
+    // Bit flags to determine borders around pixel
+    static const unsigned char NO_BORDER    = 0;      // 0000 0000 
+    static const unsigned char LEFT_BORDER  = 1 << 0; // 0000 0001 
+    static const unsigned char LOWER_BORDER = 1 << 1; // 0000 0010
+    static const unsigned char RIGHT_BORDER = 1 << 2; // 0000 0100
+    static const unsigned char UPPER_BORDER = 1 << 3; // 0000 1000
+
+    // Bit flags for marching square case
+    static const unsigned char ALL_LOW     = 0;      // 0000 0000 
+    static const unsigned char UPPER_LEFT  = 1 << 0; // 0000 0001 
+    static const unsigned char LOWER_LEFT  = 1 << 1; // 0000 0010
+    static const unsigned char LOWER_RIGHT = 1 << 2; // 0000 0100
+    static const unsigned char UPPER_RIGHT = 1 << 3; // 0000 1000
+    static const unsigned char ALL_HIGH    = UPPER_LEFT | LOWER_LEFT | LOWER_RIGHT | UPPER_RIGHT; // 0000 1111
+    static const unsigned char SADDLE_NW   = UPPER_LEFT | LOWER_RIGHT; // 0000 0101
+    static const unsigned char SADDLE_NE   = UPPER_RIGHT | LOWER_LEFT; // 0000 1010
+
+    typedef std::pair< Point, Point > Segment;
+    typedef std::pair< ValuedPoint, ValuedPoint > ValuedSegment;
+
+    //
+    // An array of segments, at most 3 segments
+    struct Segments
+    {
+        Segments(): sz(0) {}
+        Segments(const Segment &first_): sz(1), segs({first_}) {}
+        Segments(const Segment &first_, const Segment &second_): sz(2), segs({first_, second_}) {}
+        Segments(const Segment &first_, const Segment &second_, const Segment &third_): sz(3), segs({first_, second_, third_}) {}
+
+        std::size_t size() const {return sz;}
+
+        const Segment & operator[](std::size_t idx) const
+        {
+            assert(idx < sz);
+            return segs[idx];
+        }
+    private:
+        const std::size_t sz;
+        const Segment segs[3];
+    };
+
+
+    Square(const ValuedPoint & upperLeft_, const ValuedPoint & upperRight_,
+           const ValuedPoint & lowerLeft_, const ValuedPoint & lowerRight_, unsigned char borders_ = NO_BORDER, bool split_ = false)
+        : upperLeft(upperLeft_)
+        , lowerLeft(lowerLeft_)
+        , lowerRight(lowerRight_)
+        , upperRight(upperRight_)
+        , nanCount((isnan(upperLeft.value) ? 1 : 0) + (isnan(upperRight.value) ? 1 : 0)
```

Comment:
would probably be better to use std::isnan() since isnan() might not be defined in the global namespace on all platforms